### PR TITLE
move inline string ABIs to separate files and embed them

### PR DESCRIPTION
**Pull request by ceyonur**
## Why this should be merged

Moves inline ABI strings in precompiles to embedded separate string files.

## How this works

Changes how we handle `abi.Bind` function and adds a hook to provide a way for expanding the function without expanding the standard inputs. Creates inputted ABI files in the same directory with output precompile file, and embeds those ABI files in precompile.  If the output precompile file is STDOUT then it will inline ABI strings into the precompile (old behaviour).

Changes precompile template to reflect those changes. 

## How this was tested

Created a RewardManager precompile with precompile gen tool and run `stateful_precompile_test.go` tests. Also confirmed the ABI is inlined when output is STDOUT.

### CODE REVIEW

File: accounts/abi/bind/bind.go [hunk 1]
```diff
@@ -291,19 +292,14 @@ func Bind(types []string, abis []string, bytecodes []string, fsigs []map[string]
...
 
-	// Generate the contract template data according to contract type (precompile/non)
-	if isPrecompile {
-		if lang != LangGo {
-			return "", errors.New("only GoLang binding for precompiled contracts is supported yet")
-		}
-
-		if len(contracts) != 1 {
-			return "", errors.New("cannot generate more than 1 contract")
+	// Generate the contract template data according to  hook
```
Comment (darioush) [inline accounts/abi/bind/bind.go:295]:
```suggestion
	// Generate the contract template data according to hook
```

File: precompile/reward_manager.go [hunk 2]
```diff
@@ -26,9 +28,6 @@ const (
...
...ddress\",\"name\":\"rewardAddress\",\"type\":\"address\"}],\"stateMutability\":\"view\",\"type\":\"function\"},{\"inputs\":[],\"name\":\"disableRewards\",\"outputs\":[],\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"inputs\":[{\"internalType\":\"address\",\"name\":\"addr\",\"type\":\"address\"}],\"name\":\"readAllowList\",\"outputs\":[{\"internalType\":\"uint256\",\"name\":\"role\",\"type\":\"uint256\"}],\"stateMutability\":\"view\",\"type\":\"function\"},{\"inputs\":[{\"internalType\":\"address\",\"name\":\"addr\",\"type\":\"address\"}],\"name\":\"setAdmin\",\"outputs\":[],\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"inputs\":[{\"internalType\":\"address\",\"name\":\"addr\",\"type\":\"address\"}],\"name\":\"setEnabled\",\"outputs\":[],\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"inputs\":[{\"internalType\":\"address\",\"name\":\"addr\",\"type\":\"address\"}],\"name\":\"setNone\",\"outputs\":[],\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"inputs\":[{\"internalType\":\"address\",\"name\":\"addr\",\"type\":\"address\"}],\"name\":\"setRewardAddress\",\"outputs\":[],\"stateMutability\":\"nonpayable\",\"type\":\"function\"}]"
```
Comment (darioush) [inline precompile/reward_manager.go:31]: ❤️

File: accounts/abi/bind/bind.go [hunk 3]
```diff
@@ -101,7 +103,11 @@ func isKeyWord(arg string) bool {
 // to be used as is in client code, but rather as an intermediate struct which
 // enforces compile time type safety and naming convention opposed to having to
 // manually maintain hard coded strings that break on runtime.
-func Bind(types []string, abis []string, bytecodes []string, fsigs []map[string]string, pkg string, lang Lang, libs map[string]string, aliases map[string]string, isPrecompile bool) (string, error) {
+func Bind(types []string, abis []string, bytecodes []string, fsigs []map[string]string, pkg string, lang Lang, libs map[string]string, aliases map[string]string) (string, error) {
+	return BindHelper(types, abis, bytecodes, fsigs, pkg, lang, libs, aliases, nil)
+}
+
+func BindHelper(types []string, abis []string, bytecodes []string, fsigs []map[string]string, pkg string, lang Lang, libs map[string]string, aliases map[string]string, bindHook BindHook) (string, error) {
```
Comment (darioush) [inline accounts/abi/bind/bind.go:110]: does `BindHelper` need to be exported?

File: accounts/abi/bind/bind.go [hunk 4]
```diff
@@ -291,19 +292,14 @@ func Bind(types []string, abis []string, bytecodes []string, fsigs []map[string]
...
+		var err error
+		data, templateSource, err = bindHook(lang, types, contracts, structs)
+		if err != nil {
+			return "", err
 		}
-		precompileType := types[0]
-		firstContract := contracts[precompileType]
-		data, templateSource = createPrecompileDataAndTemplate(firstContract, structs)
-	} else {
+	} else { // default to generate contract binding
```
Comment (darioush) [inline accounts/abi/bind/bind.go:302]: an alternative would be to pass this code in as `bindHook` for the default (`nil`) case. however there is the downside of making it more difficult to trace upstream changes. Not suggesting we make a change, just highlighting the tradeoff

Comment (ceyonur) [reply accounts/abi/bind/bind.go:302]: yeah, true. the reason I added this hook is to trace upstream changes more easily.

File: accounts/abi/bind/bind.go (hunk 3 above)
Comment (ceyonur) [reply accounts/abi/bind/bind.go:110]: nope, good catch. unexported it.

**Review by darioush — APPROVED** [review]

**Review by aaronbuchwald — APPROVED** [review]
LGTM - can we start a `precompile` branch to batch our planned precompile improvements into, so we can do one pass of updating the documentation when they are done?

**Pull request merged**